ruff.toml

```diff
@@ -11,3 +13,12 @@ select = [
     "SIM118",  # use 'key in dict'
     "SIM2",    # simplify boolean comparisons
 ]
+
+target-version = "py38"
+
+[per-file-ignores]
+"__init__.py" = ["I"]
+"setup.py" = ["I"]
+"examples/*.py" = ["I"]
+"doc/*.py" = ["I"]
+"tools/*.py" = ["I"]
```

Comment:
what is the reason for skipping these?

Reply:
ruff (isort) will mess up the whole import structures and break the package if we let it format `__init.py__` , and we were running isort only on the `networkx` folder. So I'm skipping the other folders to make ruff work without making any changes. I think we can safely un-skip examples, docs and tools folder.

Reply:
Would it not be preferable to use per-line 'isort:skip' invocations for this use case?

Reply:
Wouldn't that mean the whole file would have `isort:skip`  (the `__init.py__` ones)? As the `__init__.py` are literally just imports all around.

Reply:
That depends. Why does sorting imports in init files in this package cause them to break?

Isort has loads of options for controlling the way it is applied in the case that a subset of imports are order-dependent

See https://pycqa.github.io/isort/docs/configuration/action_comments.html

Reply:
> That depends. Why does sorting imports in init files in this package cause them to break?

We want the user namespace to be filled in a certain way, as this requires us to follow a certain import order. Like to have the `py_random_state` decorator requires `create_py_random_state` to be imported first from utils.misc.

> Isort has loads of options for controlling the way it is applied in the case that a subset of imports are order-dependent
> See https://pycqa.github.io/isort/docs/configuration/action_comments.html

Well in that case let's just put this in a single config file, rather than go grep-ing for `# isort:` to find what we skip and what we do not. 

Let's remember none of this makes actual difference to the package itself, it's all for aesthetic changes so we don't have to think about any of this ever again :)

Reply:
> > That depends. Why does sorting imports in init files in this package cause them to break?
> 
> We want the user namespace to be filled in a certain way, as this requires us to follow a certain import order. Like to have the `py_random_state` decorator requires `create_py_random_state` to be imported first from utils.misc.

sorry, i'm struggling to understand this. Please bear with me!

i see that `utils.misc` exports a method called `create_py_random_state` which is re-exported in `utils.__init__`. The `py_random_state` decorator is then defined in `utils.decorators`. Which part of that is impacted by the order of imports in `utils.__init__`?

assuming there is genuinely an order dependency there, can it be resolved by changing the import in `utils.decorators` from

```python
...
from networkx.utils import create_py_random_state, create_random_state
```

to

```python
...
from networkx.utils.misc import create_py_random_state, create_random_state
```
?

It seems to me that the current approach invites circular dependencies a little more readily than it needs to.

Reply:
that seems fair enough. can save that for a different PR!